internal/shim-sev/src/hostmap.rs

```diff
@@ -0,0 +1,204 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Provides a mapping of the shim physical addresses
+//! to the hypervisor host virtual addresses, needed
+//! for proxying syscalls.
+
+use crate::addr::{HostVirtAddr, ShimPhysUnencryptedAddr};
+
+use crate::spin::RWLocked;
+use crate::SHIM_CAN_PRINT;
+use core::alloc::Layout;
+use core::mem::{align_of, size_of};
+use core::sync::atomic::Ordering;
+use linked_list_allocator::Heap;
+use lset::Line;
+use primordial::{Address, Page as Page4KiB};
+use spinning::{Lazy, RwLockWriteGuard};
+
+/// The global [`HostMap`](HostMap) RwLock
+pub static HOSTMAP: Lazy<RWLocked<HostMap>> =
+    Lazy::new(|| RWLocked::<HostMap>::new(HostMap::new()));
+
+struct HostMemListPageHeader {
+    next: Option<&'static mut HostMemListPage>,
+}
+
+#[derive(Clone, Copy)]
+struct HostMemEntry {
+    start: usize,
+    end: usize,
+    virt_offset: i64,
+}
+
+/// Number of memory list entries per page
+pub const HOST_MEM_LIST_NUM_ENTRIES: usize = (Page4KiB::size()
+    - core::mem::size_of::<HostMemListPageHeader>())
+    / core::mem::size_of::<HostMemEntry>();
+
+struct HostMemListPage {
+    header: HostMemListPageHeader,
+    ent: [HostMemEntry; HOST_MEM_LIST_NUM_ENTRIES],
+}
+
+/// A mapping of the shim physical addresses to the hypervisor host virtual addresses
+pub struct HostMap {
+    num_pages: usize,
+    host_mem: HostMemListPage,
+}
+
+impl HostMap {
+    #[allow(clippy::integer_arithmetic)]
+    fn new() -> Self {
+        HostMap {
+            num_pages: 0,
+            host_mem: HostMemListPage {
+                header: HostMemListPageHeader { next: None },
+                ent: [HostMemEntry {
+                    start: 0,
+                    end: 0,
+                    virt_offset: 0,
+                }; HOST_MEM_LIST_NUM_ENTRIES],
+            },
+        }
+    }
+
+    fn get_virt_offset(&self, addr: usize) -> Option<i64> {
+        let mut free = &self.host_mem;
+        loop {
+            for i in free.ent.iter() {
+                if i.start == 0 {
```

Comment:
Trying to wrap my head around how all of this works together as I haven't looked at the virtual address components at all. When does the `HostMemEntry.start` get updated to hold a non-zero value after `HostMap::new()`? Is there an example I can see of how this part will be used? I get why it should be a linked list and how `NUM_ENTRIES` is determined but I think I'm missing something about what the data in `HostMemEntry` means.

Reply:
Changed the logic to recognize `end == 0` as an uninitialized element, because it is that's how the initial memory contents is set to with `page.write_bytes(0, 1)`.

Added more comments in the hope it clarifies the logic.